### PR TITLE
Vs add popups

### PR DESCRIPTION
**Pull request by Vadser**

### CODE REVIEW

File: app/assets/javascripts/categories_popups.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        modal: true,
+        open: function(){
+          fetchCategoryFormContent(id);
+        },
+        close: function(){
+          $(this).dialog('destroy');
+          $(this).remove();     
+        }
+    });
+    var $dialog = $(".ui-dialog");
```
Comment (SashaShostyr) [inline app/assets/javascripts/categories_popups.js]: use function addStylesToDialog() from dialog.js

**Review by SashaShostyr — APPROVED** [review]

File: app/assets/javascripts/categories_popups.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    $('#category_form').dialog('destroy');
+    $('#category_form').remove();
+  });
+
+  $(document).on('click','form#save_category button[type="submit"]', function(e) {
+    e.preventDefault();
+    var form = $(this).parents('form#save_category');
+    var data = $(form).serialize();
+    if (!validateCategoryForm(form)) {
+      $('#error_explanation').addClass('alert alert-warning').html('Please input name for category!').fadeIn().fadeOut(10000);
```
Comment (anatoliliotych) [inline app/assets/javascripts/categories_popups.js]: I see code duplication

File: app/assets/javascripts/categories_popups.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      url: $(form).attr('action'),
+      data: data,
+      dataType: "JSON"
+    }).success(function(data){
+      if (data.status == 'ok') {
+        $('#category_form').dialog('destroy');
+        $('#category_form').remove();
+        window.location = '/categories';
+      }else{
+        $('#error_explanation').addClass('alert alert-warning').html('Category cannot be a descendant of itself!').fadeIn().fadeOut(10000);
```
Comment (anatoliliotych) [inline app/assets/javascripts/categories_popups.js]: move to function

File: app/assets/javascripts/categories_popups.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  function validateCategoryForm(form){
+    var name_input = $(form).find('#category_name');
+    return name_input.val() != '';
+  }
+
+  function fetchCategoryFormContent(id){
+    $.ajax({
+      url: '/categories/new',
+      data: { id: id},
```
Comment (anatoliliotych) [inline app/assets/javascripts/categories_popups.js]: style

File: app/assets/javascripts/categories_popups.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      }else{
+        $('#error_explanation').addClass('alert alert-warning').html('Category cannot be a descendant of itself!').fadeIn().fadeOut(10000);
+      }
+    });
+    return false;
+  });
+
+  function validateCategoryForm(form){
+    var name_input = $(form).find('#category_name');
+    return name_input.val() != '';
```
Comment (anatoliliotych) [inline app/assets/javascripts/categories_popups.js]: check length

File: app/assets/javascripts/categories_popups.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      success: function(data){
+        $('#category_form').html(data);
+      }
+    });
+  }
+
+  function fetchUpdateCategoryContent(id){
+    $.ajax({
+      url: '/categories/' + id + '/edit',
+      data: { id: id},
```
Comment (anatoliliotych) [inline app/assets/javascripts/categories_popups.js]: style

File: app/assets/javascripts/nishes_popups.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        modal: true,
+        open: function(){
+          fetchUpdateNicheContent(id);
+        },
+        close: function(){
+          $(this).dialog('destroy');
+          $(this).remove();     
+        }
+    });
+    var $dialog = $(".ui-dialog");
```
Comment (anatoliliotych) [inline app/assets/javascripts/nishes_popups.js]: you've missed this!!!

File: app/assets/javascripts/nishes_popups.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    $('#niche_form').dialog('destroy');
+    $('#niche_form').remove();
+  });
+
+  $(document).on('click','form#save_niche button[type="submit"]', function(e) {
+    e.preventDefault();
+    var form = $(this).parents('form#save_niche');
+    var data = $(form).serialize();
+    if (!validateNicheForm(form)) {
+      $('#error_explanation').addClass('alert alert-warning').html('Please input name for niche!').fadeIn().fadeOut(10000);
```
Comment (anatoliliotych) [inline app/assets/javascripts/nishes_popups.js]: move to function

File: app/assets/javascripts/nishes_popups.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      url: $(form).attr('action'),
+      data: data,
+      dataType: "JSON"
+    }).success(function(data){
+      if (data.status == 'ok') {
+        $('#niche_form').dialog('destroy');
+        $('#niche_form').remove();
+        window.location = '/niches';
+      }else{
+        $('#error_explanation').addClass('alert alert-warning').html('Niche cannot be a descendant of itself!').fadeIn().fadeOut(10000);
```
Comment (anatoliliotych) [inline app/assets/javascripts/nishes_popups.js]: code duplication

File: app/assets/javascripts/nishes_popups.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+  function validateNicheForm(form){
+    var name_input = $(form).find('#niche_name');
+    return name_input.val() != '';
+  }
+
+  function fetchNicheFormContent(id){
+    $.ajax({
+      url: '/niches/new',
+      data: { id: id},
```
Comment (anatoliliotych) [inline app/assets/javascripts/nishes_popups.js]: style

File: app/assets/javascripts/nishes_popups.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+      success: function(data){
+        $('#niche_form').html(data);
+      }
+    });
+  }
+
+  function fetchUpdateNicheContent(id){
+    $.ajax({
+      url: '/niches/' + id + '/edit',
+      data: { id: id},
```
Comment (anatoliliotych) [inline app/assets/javascripts/nishes_popups.js]: style

File: app/views/categories/_form.html.slim [hunk 12]
```diff
@@ -1,18 +1,20 @@
...
-    = form.label :parent_id
-    = form.collection_select(:parent_id, current_user.categories, :id, :name, :include_blank => "Please select")
-  .actions
-    = form.submit
+.card
+  .header
+    h2
+      | Category
+  .body
+    = form_for(@category, :html=>{ :id=>'save_category', :multipart => true}) do |form|
```
Comment (anatoliliotych) [inline app/views/categories/_form.html.slim]: style

File: app/views/categories/_form.html.slim [hunk 13]
```diff
@@ -1,18 +1,20 @@
...
+          = form.text_field :name, class: 'form-control'
+      = form.label :parent
+      .form-group
+        .form-line  
+          = form.collection_select :parent_id,  current_user.categories, :id, :name, include_blank: true, class: 'form-control'
+      br
+      .button-demo
+        button#close.btn.btn-danger.waves-effect
+          | Cancel    
+        = form.button 'Save', type: :submit, class: 'btn btn-primary waves-effect'
```
Comment (anatoliliotych) [inline app/views/categories/_form.html.slim]: empty line symbol

File: app/views/niches/_form.html.slim [hunk 14]
```diff
@@ -1,18 +1,20 @@
...
-    = form.label :parent
-    = form.collection_select :parent_id, current_user.niches, :id, :name, include_blank: true
-  .actions
-     = form.submit
+.card
+  .header
+    h2
+      | Niche
+  .body
+    = form_for(@niche, :html=>{ :id=>'save_niche', :multipart => true}) do |form|
```
Comment (anatoliliotych) [inline app/views/niches/_form.html.slim]: style

File: spec/controllers/categories_controller_spec.rb [hunk 15]
```diff
@@ -90,9 +90,7 @@
       it 'redirects to the created category' do
         post :create, params: { category: valid_attributes }
         category = assigns(:category)
-        expect(response).to have_http_status(:redirect)
-        expect(response).to redirect_to(category)
-        expect(flash[:notice]).to match(/Category was successfully created\./)
+        expect(response.body).to eq("{\"status\":\"ok\"}")
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: you can parse body as json and check hashes

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/nishes_popups.js [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    });
+    return false;
+  });
+
+  function nicheStatusCheck(data){
+    if (data.status == 'ok') {
+      $('#niche_form').dialog('destroy');
+      $('#niche_form').remove();
+      window.location = '/niches';
+    }else{
```
Comment (anatoliliotych) [inline app/assets/javascripts/nishes_popups.js]: spaces

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

**Review by anatoliliotych — APPROVED** [review]

**Pull request merged**